apollo-federation/src/connectors/json_selection/apply_to.rs

```diff
@@ -661,95 +718,105 @@ impl ApplyToInternal for WithRange<PathList> {
-                            if shape.is_none() {
-                                shape.clone()
-                            } else {
-                                rest.compute_output_shape(
-                                    field(shape, key, source_id),
-                                    dollar_shape.clone(),
-                                    named_var_shapes,
-                                    source_id,
-                                )
-                            }
-                        })
-                        .collect::<Vec<_>>();
-
-                    let mapped_rest = if tail.is_none() {
-                        tail.clone()
-                    } else {
-                        rest.compute_output_shape(
-                            field(tail, key, source_id),
-                            dollar_shape,
-                            named_var_shapes,
-                            source_id,
-                        )
-                    };
-
-                    Shape::array(mapped_prefix, mapped_rest, input_shape.locations)
-                } else {
-                    rest.compute_output_shape(
-                        field(&input_shape, key, source_id),
-                        dollar_shape,
-                        named_var_shapes,
-                        source_id,
-                    )
-                }
+                (child_shape, Some(tail))
             }
 
-            PathList::Expr(expr, tail) => tail.compute_output_shape(
+            PathList::Expr(expr, tail) => (
                 expr.compute_output_shape(
                     input_shape,
                     dollar_shape.clone(),
                     named_var_shapes,
                     source_id,
                 ),
-                dollar_shape,
-                named_var_shapes,
-                source_id,
+                Some(tail),
             ),
 
-            PathList::Method(method_name, _method_args, _tail) => ArrowMethod::lookup(method_name)
-                .map_or_else(
-                    || {
+            PathList::Method(method_name, method_args, tail) => {
+                if let Some(method) = ArrowMethod::lookup(method_name) {
+                    (
+                        method.shape(
+                            method_name,
+                            method_args.as_ref(),
+                            input_shape,
+                            dollar_shape.clone(),
+                            named_var_shapes,
+                            source_id,
+                        ),
+                        Some(tail),
+                    )
+                } else {
+                    (
                         Shape::error(
                             format!("Method ->{} not found", method_name.as_str()),
                             method_name.shape_location(source_id),
-                        )
-                    },
-                    |_method| Shape::unknown(method_name.shape_location(source_id)),
```

Comment:
Here's where we were previously returning the `Unknown` shape for all `->` methods.